plugins/arm_a64/a64_defs.ml

```diff
@@ -0,0 +1,213 @@
+open Core_kernel
+open Regular.Std
+open Bap.Std
+
+module QWord = Bitvec.M64
+module Word = Bitvec.M32
+module Byte = Bitvec.M8
+
+type registers_t = [
+  (* parameter and result registers *)
+  | `X0 | `X1 | `X2 | `X3 | `X4 | `X5 | `X6 | `X7 
+
+  (* caller saved temporary registers *)
+  | `X9 | `X10 | `X11 | `X12 | `X13 | `X14 | `X15 
+
+  (* callee saved registers *) 
+  | `X19 | `X20 | `X21 | `X22 | `X23 | `X24 | `X25 | `X26 | `X27 | `X28
+
+  | `X8
+  | `X16 | `X17 | `X18
+  | `X29 | `X30
+
+  (* parameter and result registers *)
+  | `W0 | `W1 | `W2 | `W3 | `W4 | `W5 | `W6 | `W7 
+
+  (* caller saved temporary registers *)
+  | `W9 | `W10 | `W11 | `W12 | `W13 | `W14 | `W15 
+
+  (* callee saved registers *) 
+  | `W19 | `W20 | `W21 | `W22 | `W23 | `W24 | `W25 | `W26 | `W27 | `W28
+
+  | `W8
+  | `W16 | `W17 | `W18
+  | `W29 | `W30
+
+  | `WZR (* 32-bit zero register *)
+  | `XZR (* 64-bit zero register *)
+  | `WSP (* 32-bit stack pointer *)
+  | `SP (* 64-bit stack pointer *)
+  | `PC (* program pointer *)
+  | `ELR (* exception link register *)
+  | `SPSR (* saved processor state register *)
+]
+
+type sign_t = Signed | Unsigned
+type operation_t = LD | ST
+type bv_size_t = 
+  | B (* byte, 8 bits *) 
+  | H (* halfword, 16 bits *)
+  | W (* word, 32 bits *)
+  | X (* doubleword, 64 bits *)
+
+type extend_t =
+  | SXTB
+  | SXTH
+  | SXTW
+  | SXTX
+  | UXTB
+  | UXTH
+  | UXTW
+  | UXTX
+  | LSL
+
+type conditon_t = [
+  | `EQ
+  | `NE
+  | `CS | `HS
+  | `CC | `LO
+  | `MI
+  | `PL
+  | `VS
+  | `VC
+  | `HI
+  | `LS
+  | `GE
+  | `LT
+  | `GT
+  | `LE
+  | `AL
+  | `NV
+]
+
+type operand_t = [
+  | `Reg of registers_t
+  | `Imm of Bitvec.t
+]
+
+type branch_insn_t = [
+  (* conditional branch TODO*)
+  | `CBNZ
+  (* unconditional branch (immediate) *)
+  | `B
+  | `BL
+  (* unconditional branch (register) *)
+  | `BLR
+  | `BR
+  | `RET
+]
+
+type mem_access_insn_t = [
+  (* load-store single register TODO *)
```

Comment:
**Question:**
There are a lot of variants for one single instruction, eg. the following three segments are all variant of `LDR`, and I need to distinguish them. 

Now, I define their type at here(very ugly), and lift the specific variant using more general functions here [https://github.com/BinaryAnalysisPlatform/bap/pull/1141/files#diff-6b5b4bf2eb12277cf6bf719ac07f9bd3R20](url).  

Am I on the right way? If yes, any suggestions?

Reply:
It will apply only to LDR and STR instructions, while the rest are less complex. I think it shouldn't be a problem. I don't know what @ivg think of that.

Reply:
I provided a big review and suggest starting from scratch, I think we have a false start. See the comment below.

Reply:
Hi @ivg , just read your comments. Thank you for the suggestions!

 Is [extend_reg](https://github.com/BinaryAnalysisPlatform/bap/blob/1d848484fc20f0d07202804a02e9eedb028d6cba/plugins/arm_a64/a64_utils.ml#L92) the "small and easy to use function" you commented (although it may be a bad implementation)? They are all in part J(Armv8 Pseudocode) of the official manual. Once we have these small functions, we still need to implement pseudo-code in the specification(actually it is what I am trying to do, I wrote these code following those pseudo-code). Thus, if I understand what you mean correctly, your suggestion is actually let me stop merging all load variants into one function. Instead, write the ocaml/core theory version of the pseudo-code for every instruction variant(one ocaml function per instruction variant). 

Please correct me if I am wrong.

Reply:
Yep, it is close to be "small and easy to use") See my comment for this function. 

> Thus, if I understand what you mean correctly, your suggestion is actually let me stop merging all load variants into one function. Instead, write the ocaml/core theory version of the pseudo-code for every instruction variant(one ocaml function per instruction variant).

Exactly and perfectly worded! As close to the spec as possible. 

Also, another advice, try not to correct the spec and follow it, no matter how it looks. You might find yourself thinking, why is this expressed so ugly, we can write it nicely using this core theory operation, or this OCaml construct. And while it is in general a good idea when you write a program in our case we should refrain ourselves from doing this, as our task to express the semantics as close as possible in our meta language.